Hud driver conversion

BrettASwift opened the pull request:
@kadst43 @AmbientOSX @nmarshak1337 

This is complete. All of the HUD elements have been converted, except for the Blockly trace element which we discussed.

Requires: https://github.com/virtual-world-framework/vwf/pull/466

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+          }
+        }
+      }
+      context.globalAlpha = alpha;
```
Is there a reason to have the code in yaml?  Wherever possible, I'd rather have it in a .js file...

BrettASwift replied:
Sort of. It helps when writing the draw functions to be able to see all of the properties available to you. The draw functions, by the way, are used only by the HUD driver. They're not true methods. I would like to find a way to make that clear, but I have not yet found a way. `this` in the draw function refers to the view object (HUD element) rather than the VWF node itself.

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+      - context
+      - element
+    body: |
+      context.globalAlpha = 1;
```
Is this not using the element parameter?

BrettASwift replied:
Nope. That's just there because the element is being passed in under the covers. It's more for our sake, just to know it's there if we need it. As per my other comment, I need to find a way to make some things more clear to the developer.

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+  setAllBlinking:
+  setAllEnabled:
+children:
+  blocklyButton:
```
Given that the children are moderately complex, would it make sense to give each child its own .js and .yaml files, maybe all in some subdirectory like source/hud (so we don't clutter up the source directory too much)?  I don't feel strongly for the HUD, but it seems like that might be a bit cleaner.

BrettASwift replied:
I suppose it won't be too much trouble.

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+        body: |
+          if ( this.icon ) {
+            context.drawImage( this.icon, position.x, position.y );
+          }
```
Code in a .js file?

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+        body: |
+          if ( this.icon ) {
+            context.drawImage( this.icon, position.x, position.y );
+          }
```
code in .js?

kadst43 commented on source/marsGameHud.vwf.yaml:
```diff
@@ -0,0 +1,468 @@
+          }
+          var images = this.images;
+          images.activeModeIcon = src;
+          this.images = images;
```
With a setter this complex, I would add a method named something like setActiveMode$, define it in the .js, and then just call it here.

BrettASwift commented:
I'll break up everything into their own files and push it up.

BrettASwift commented:
Moved the elements to their own files, added license info, etc. Had to update the driver to handle moving the draw functions into external scripts as well.

kadst43 commented on source/scenario/scenario1a.vwf.yaml:
```diff
@@ -103,15 +103,15 @@ children:
           - setObjective:
             - "Click the PROGRAM button to open your control processor."
-          - setHUDProperty:
-            - camera_topDown
```
Can we get rid of the setHUDProperty action?  Any other actions we can get rid of/simplify now?

kadst43 commented:
I like this a lot better - much cleaner. :)

Aside from the one comment (and I assume that you've thoroughly tested... ;) ): :+1:
